File: golang/linked_list/lru_cache.go
package main

type LRUCache struct {
	capacity int
	cache    map[int]*DoublyLinkedListNode
	head     *DoublyLinkedListNode
	tail     *DoublyLinkedListNode
}

type DoublyLinkedListNode struct {
	key, value int
	prev, next *DoublyLinkedListNode
}

func Constructor(capacity int) LRUCache {
	head := &DoublyLinkedListNode{}
	tail := &DoublyLinkedListNode{}
	head.next = tail
	tail.prev = head

	return LRUCache{
		capacity: capacity,
		cache:    make(map[int]*DoublyLinkedListNode),
		head:     head,
		tail:     tail,
	}
}

func (this *LRUCache) Get(key int) int {
	if node, found := this.cache[key]; found {
		this.moveToFront(node)
		return node.value
	}
	return -1
}

func (this *LRUCache) Put(key int, value int) {
	if node, found := this.cache[key]; found {
		node.value = value
		this.moveToFront(node)
	} else {
		if len(this.cache) >= this.capacity {
			this.removeLRU()
		}
		node := &DoublyLinkedListNode{key: key, value: value}
		this.cache[key] = node
		this.addToFront(node)
	}
}

func (this *LRUCache) moveToFront(node *DoublyLinkedListNode) {
	this.removeNode(node)
	this.addToFront(node)
}

func (this *LRUCache) addToFront(node *DoublyLinkedListNode) {
	node.next = this.head.next
	node.prev = this.head
	this.head.next.prev = node
	this.head.next = node
}

func (this *LRUCache) removeNode(node *DoublyLinkedListNode) {
	node.prev.next = node.next
	node.next.prev = node.prev
}

func (this *LRUCache) removeLRU() {
	lru := this.tail.prev
	this.removeNode(lru)
	delete(this.cache, lru.key)
}

File: java/src/main/java/com/sanedge/interview/linked_list/LRUCache.java
package com.sanedge.interview.linked_list;

import java.util.HashMap;
import java.util.Map;

class DoublyLinkedListNode {
  int key, value;
  DoublyLinkedListNode prev, next;

  public DoublyLinkedListNode() {
  }

  public DoublyLinkedListNode(int key, int value) {
    this.key = key;
    this.value = value;
  }
}

public class LRUCache {
  private int capacity;
  private Map<Integer, DoublyLinkedListNode> cache;
  private DoublyLinkedListNode head, tail;

  public LRUCache(int capacity) {
    this.capacity = capacity;
    this.cache = new HashMap<>();
    this.head = new DoublyLinkedListNode();
    this.tail = new DoublyLinkedListNode();
    head.next = tail;
    tail.prev = head;
  }

  public int get(int key) {
    if (cache.containsKey(key)) {
      DoublyLinkedListNode node = cache.get(key);
      moveToFront(node);
      return node.value;
    }

    return -1;
  }

  public void put(int key, int value) {
    if (cache.containsKey(key)) {
      DoublyLinkedListNode node = cache.get(key);
      node.value = value;
      moveToFront(node);
    } else {
      if (cache.size() >= capacity) {
        removeLRU();
      }
      DoublyLinkedListNode newNode = new DoublyLinkedListNode(key, value);
      cache.put(key, newNode);
      addToFront(newNode);
    }
  }

  private void moveToFront(DoublyLinkedListNode node) {
    removeNode(node);
    addToFront(node);
  }

  private void addToFront(DoublyLinkedListNode node) {
    node.next = head.next;
    node.prev = head;
    head.next.prev = node;
    node.next = node;
  }

  private void removeNode(DoublyLinkedListNode node) {
    node.prev.next = node.next;
    node.next.prev = node.prev;
  }

  private void removeLRU() {
    DoublyLinkedListNode lru = tail.prev;
    removeNode(lru);
    cache.remove(lru.key);
  }

}
